test/apps/mandelbrot.c

```diff
@@ -0,0 +1,507 @@
+
+        // Check if all work at this PE has been done
+        if (work_start >= pe_ct_max[rr_pe]) { // No more work at this PE
+            pe_pending--;
+            pe_mask[rr_pe] = 0;
+            continue;
+        }
+
+        if (work_end >= pe_ct_max[rr_pe]) { // This is the last bit of work on this PE
+            work_end = pe_ct_max[rr_pe];
+            pe_pending--;
+            pe_mask[rr_pe] = 0;
+        }
+
+        if (!use_blocking)
+            shmem_ctx_quiet(ctx[index]);
+
+        // Do actual compute work
+        for (i = work_start, j = 0; i < work_end; i++, j++) {
+            pixels[index][j] = computeSingle(i%width, i/width);
+        }
+
+        // Return the computed image data to the PE responsible for it
+        if (use_blocking)
+            shmem_ctx_putmem(ctx[index], &imageData[work_start], pixels[index],
+                             (work_end-work_start)*sizeof(int), rr_pe);
+        else
+            shmem_ctx_putmem_nbi(ctx[index], &imageData[work_start], pixels[index],
+                                 (work_end-work_start)*sizeof(int), rr_pe);
+
+        total_work += work_end - work_start;
+
+        if (use_pipelining)
+            index ^= 1;
+    }
+    shmem_quiet();
+    timer = getTime() - timer;
+
+    // send stats data to IMAGE_PE
+    shmem_long_add(&sumTime, timer, IMAGE_PE);
+    shmem_long_add(&sumWorkRate, (long)(total_work / ((double)timer / 1e6) + 0.5), IMAGE_PE);
+
+#ifdef ENABLE_PAPI
+    // Read PAPI cache miss counters
+    PAPI_read_counters(counters, 4);
+    shmem_long_add(&sumL1_ICM, counters[0]*1e6/counters[2], IMAGE_PE);
+    shmem_long_add(&sumL2_ICM, counters[1]*1e6/counters[3], IMAGE_PE);
+#endif
+
+    free(pixels[0]);
+    free(pixels[1]);
+    free(pe_mask);
+    free(pe_ct_max);
+
+    return NULL;
+}
+
+static void printUsage() {
+    printf("USAGE: mandelbrot [options]\n");
+    printf("                  -t <num_threads> number of worker threads (def: 1)\n");
+    printf("                  -w <width>       width of the mandelbrot domain (def: 2048)\n");
+    printf("                  -w <height>      height of the mandelbrot domain (def: 2048)\n");
+    printf("                  -j <job_points>  load balancing granularity (def: 128)\n");
+    printf("                  -o               output image mandelbrot.pgm (def: off)\n");
+    printf("                  -c               use OpenSHMEM contexts (def: on)\n");
+    printf("                  -p               enable pipelining (implies -c) (def: on)\n");
+    printf("                  -b               use blocking communication (def: off)\n");
+    printf("                  -?               prints this message\n");
+}
+
+int main(int argc, char** argv) {
+    int tl, i;
+    int c;
+    int num_threads = 1;
+    int out_file = 0;
+    pthread_t *threads;
+    struct th_arg *t_arg;
+
+#ifdef ENABLE_PINNING
+    int p4_cpu;
+    pthread_t thread = pthread_self();
+    cpu_set_t cpuset;
+#endif
+
```

Comment:
I'd prefer to leave this as-is, since there are different ways of achieving the pinning (e.g. via ```mpiexec``` or explicitly via ```pthreads_setaffinity```) and the right solution may depend on whether threads are spawned lower down in the networking stack.